### PR TITLE
Vectorize `basic_string::find_first_of`

### LINKED ISSUE #4497: Further optimize `find_first_of`
* [x] For one-element needle use `find`
* [x] For more than 16 bytes needle check it by parts
* [x] Do something for 4 and 8 byte size elements
* [ ] Consider enabling for some cases `basic_string::find_first_of` (or prove it is not possible)

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Resolves #4497

Benchmark                                           |    main |    this 
----------------------------------------------------|---------|---------
bm<AlgType::str_member, char>/2/3                   | 5.52 ns | 5.84 ns 
bm<AlgType::str_member, char>/7/4                   | 12.7 ns | 12.7 ns 
bm<AlgType::str_member, char>/9/3                   | 13.1 ns | 11.4 ns 
bm<AlgType::str_member, char>/22/5                  | 16.9 ns | 11.9 ns 
bm<AlgType::str_member, char>/58/2                  | 30.9 ns | 13.1 ns 
bm<AlgType::str_member, char>/102/4                 | 44.8 ns | 15.6 ns 
bm<AlgType::str_member, char>/325/1                 |  101 ns | 36.0 ns 
bm<AlgType::str_member, char>/1011/11               |  269 ns |  100 ns 
bm<AlgType::str_member, char>/1502/23               |  400 ns |  316 ns 
bm<AlgType::str_member, char>/3056/7                |  763 ns |  286 ns 
bm<AlgType::str_member, wchar_t>/2/3                | 6.11 ns | 14.2 ns 
bm<AlgType::str_member, wchar_t>/7/4                | 12.2 ns | 16.2 ns 
bm<AlgType::str_member, wchar_t>/9/3                | 12.6 ns | 14.6 ns 
bm<AlgType::str_member, wchar_t>/22/5               | 17.3 ns | 15.7 ns 
bm<AlgType::str_member, wchar_t>/58/2               | 36.1 ns | 20.0 ns 
bm<AlgType::str_member, wchar_t>/102/4              | 55.7 ns | 27.2 ns 
bm<AlgType::str_member, wchar_t>/325/1              |  137 ns | 66.9 ns 
bm<AlgType::str_member, wchar_t>/1011/11            |  392 ns |  400 ns 
bm<AlgType::str_member, wchar_t>/1502/23            |  576 ns |  587 ns 
bm<AlgType::str_member, wchar_t>/3056/7             | 1142 ns |  574 ns 
bm<AlgType::str_member, wchar_t, L'\x03B1'>/2/3     | 7.12 ns | 15.9 ns 
bm<AlgType::str_member, wchar_t, L'\x03B1'>/7/4     | 24.3 ns | 27.2 ns 
bm<AlgType::str_member, wchar_t, L'\x03B1'>/9/3     | 26.3 ns | 14.6 ns 
bm<AlgType::str_member, wchar_t, L'\x03B1'>/22/5    | 63.3 ns | 15.7 ns 
bm<AlgType::str_member, wchar_t, L'\x03B1'>/58/2    | 61.3 ns | 19.5 ns 
bm<AlgType::str_member, wchar_t, L'\x03B1'>/102/4   |  134 ns | 26.9 ns 
bm<AlgType::str_member, wchar_t, L'\x03B1'>/325/1   |  141 ns | 66.2 ns 
bm<AlgType::str_member, wchar_t, L'\x03B1'>/1011/11 | 2953 ns |  429 ns 
bm<AlgType::str_member, wchar_t, L'\x03B1'>/1502/23 | 9272 ns |  908 ns 
bm<AlgType::str_member, wchar_t, L'\x03B1'>/3056/7  | 5572 ns |  560 ns

### CODE REVIEW

File: tests/std/tests/VSO_0000000_vector_algorithms/test.cpp [hunk 1]
```diff
@@ -1024,6 +1023,61 @@ void test_bitset(mt19937_64& gen) {
...
+    }
+}
+
+template <class T>
+void test_basic_string(mt19937_64& gen) {
+    using dis_int_type = conditional_t<is_signed_v<T>, int32_t, uint32_t>;
+
+    uniform_int_distribution<dis_int_type> dis_latin('a', 'z');
+    test_basic_string_dis<T>(gen, dis_latin);
+    if constexpr (sizeof(T) >= 2) {
```
Comment (StephanTLavavej) [inline tests/std/tests/VSO_0000000_vector_algorithms/test.cpp:1065]:
```
D:\a\_work\1\s\tests\std\tests\VSO_0000000_vector_algorithms\test.cpp(1065): error C2220: the following warning is treated as an error
D:\a\_work\1\s\tests\std\tests\VSO_0000000_vector_algorithms\test.cpp(1065): warning C4984: 'if constexpr' is a C++17 language extension
```

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/__msvc_string_view.hpp [hunk 2]
```diff
@@ -701,9 +701,38 @@ constexpr size_t _Traits_find_first_of(_In_reads_(_Hay_size) const _Traits_ptr_t
     // in [_Haystack, _Haystack + _Hay_size), look for one of [_Needle, _Needle + _Needle_size), at/after _Start_at
     if (_Needle_size != 0 && _Start_at < _Hay_size) { // room for match, look for it
         if constexpr (_Special) {
+#if _USE_STD_VECTOR_ALGORITHMS
+            bool _Try_vectorize = !_STD _Is_constant_evaluated() && _Hay_size - _Start_at > _Threshold_find_first_of;
+            constexpr size_t _Elem_size = sizeof(*_Haystack);
+
+            // Additional condition for the case where the table lookup outperforms the vectorization
+            if (_Try_vectorize && (_Elem_size == 1 || _Elem_size * _Needle_size <= 16)) {
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/inc/__msvc_string_view.hpp]: Yes it was reversed

File: stl/inc/__msvc_string_view.hpp [hunk 3]
```diff
@@ -701,9 +701,38 @@ constexpr size_t _Traits_find_first_of(_In_reads_(_Hay_size) const _Traits_ptr_t
...
+                    const _Traits_ptr_t<_Traits> _Found = _STD _Find_first_of_vectorized(
+                        _Haystack + _Start_at, _Haystack + _Hay_size, _Needle, _Needle + _Needle_size);
+
+                    if (_Found != _Haystack + _Hay_size) {
+                        return static_cast<size_t>(_Found - _Haystack);
+                    } else {
+                        return static_cast<size_t>(-1); // no match
+                    }
+                }
+#endif // _USE_STD_VECTOR_ALGORITHMS
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/inc/__msvc_string_view.hpp]: The other three in `find_meow_of` family still use the complicated control flow.
I want to try to vectorize them too though, one by one.

[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/__msvc_string_view.hpp]: I want to create a cleanup PR after landing your first PR to deal with the existing control flow, making your later vectorizations easier.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by AlexGuteniev** [comment]
My experiment on having `_Use_bitmap` to always true or always false

Benchmark                                |never bitmap |always bitmap       
-----------------------------------------|-------------|---------------
bm<AlgType::str_member, char>/22/5       | **11.3 ns**     |   17.5 ns     
bm<AlgType::str_member, char>/58/2       | **12.6 ns**     |   41.2 ns     
bm<AlgType::str_member, char>/102/4      | **15.4 ns**     |   54.8 ns     
bm<AlgType::str_member, char>/325/1      | **35.7 ns**     |    132 ns     
bm<AlgType::str_member, char>/1011/11    | **96.8 ns**     |    355 ns     
bm<AlgType::str_member, char>/1502/23    |  **320 ns**     |    607 ns     
bm<AlgType::str_member, char>/3056/7     |  **282 ns**     |   1026 ns     
bm<AlgType::str_member, wchar_t>/22/5    | **12.5 ns**     |   17.8 ns     
bm<AlgType::str_member, wchar_t>/58/2    | **16.9 ns**     |   31.5 ns     
bm<AlgType::str_member, wchar_t>/102/4   | **23.4 ns**     |   54.6 ns     
bm<AlgType::str_member, wchar_t>/325/1   | **63.0 ns**     |    134 ns     
bm<AlgType::str_member, wchar_t>/1011/11 |  406 ns     |    376 ns    
bm<AlgType::str_member, wchar_t>/1502/23 |  870 ns     |    **553 ns**     
bm<AlgType::str_member, wchar_t>/3056/7  |  **514 ns**     |   1100 ns     
bm<AlgType::str_member, char32_t>/22/5   | **10.7 ns**     |   21.3 ns     
bm<AlgType::str_member, char32_t>/58/2   | **9.36 ns**     |   43.0 ns     
bm<AlgType::str_member, char32_t>/102/4  | **16.7 ns**     |   58.9 ns     
bm<AlgType::str_member, char32_t>/325/1  | **18.8 ns**     |    134 ns     
bm<AlgType::str_member, char32_t>/1011/11|  338 ns     |    385 ns     
bm<AlgType::str_member, char32_t>/1502/23| 1015 ns     |    **639 ns**     
bm<AlgType::str_member, char32_t>/3056/7 |  **541 ns**     |   1109 ns

**Comment by AlexGuteniev** [comment]
## ⚠️ I suspect this is broken for custom character types ⚠️ 

I had the assumption that `_Special` is the case where we definitely have one of the standard character types. But apparently `char_traits` can be specialized for custom character types, so `_Special` may be `true` for something not behaving as a character.

**Comment by frederick-vs-ja** [comment]
> I had the assumption that `_Special` is the case where we definitely have one of the standard character types. But apparently `char_traits` can be specialized for custom character types, so `_Special` may be `true` for something not behaving as a character.

It seems that we should fix `_Special` for all `_Traits_meow` algorithms. Perhaps in another PR?
(Also, it's possibly better to make `_Special` not a template parameter, which would shorten mangled names.)

The stragety may be like this:
```C++
template <class>
constexpr bool _Is_std_char_traits_specialization = false;
template <class _CharT>
constexpr bool _Is_std_char_traits_specialization<char_traits<_CharT>> = _Is_EcharT<_CharT>;
```

**Comment by AlexGuteniev** [comment]
> It seems that we should fix `_Special` for all `_Traits_meow` algorithms. Perhaps in another PR?

Yes, I think I can do this some time later (If you or someone else won't do it before me).
I'd prefer keeping the current PR draft, until it can be merged with `main` containing the changes of that PR.
This way `basic_string` won't become more broken for odd character types.

> (Also, it's possibly better to make `_Special` not a template parameter, which would shorten mangled names.)

Good idea.

> The strategy may be like this:
Also agree.

**Comment by AlexGuteniev** [comment]
I'm marking this PR as ready again.

Turns out that `basic_string::find_meow_of` rejects unicorns here:

https://github.com/microsoft/STL/blob/91e425587a0b9b5ac0d1005844c661fee2e2813b/stl/inc/__msvc_string_view.hpp#L678-L680

Since it is compile-time error already, vectorization doesn't make this case any more broken.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
:yarn: :black_cat: :tada:
===

**Pull request merged**